core/src/main/java/com/digitalpebble/stormcrawler/protocol/file/FileResponse.java

```diff
@@ -0,0 +1,193 @@
+import java.util.Date;
+import java.util.Locale;
+
+import org.apache.http.HttpStatus;
+import org.slf4j.LoggerFactory;
+
+import com.digitalpebble.stormcrawler.Metadata;
+import com.digitalpebble.stormcrawler.protocol.HttpHeaders;
+import com.digitalpebble.stormcrawler.protocol.ProtocolResponse;
+
+public class FileResponse {
+
+    /**
+     * Adopted from Apache Nutch File Protocol implementation
+     * 
+     * @see https://github.com/apache/nutch/blob/master/src/plugin/protocol-file
+     *      /src/java/org/apache/nutch/protocol/file/FileResponse.java
+     * @see https://github.com/apache/nutch/blob/master/src/plugin/protocol-file
+     *      /src/java/org/apache/nutch/protocol/file/File.java
+     * 
+     **/
+
+    private byte[] content;
+    private int statusCode;
+    private Metadata metadata;
+    private final FileProtocol fileProtocol;
+    static final SimpleDateFormat dateFormat = new SimpleDateFormat(
+            "EEE, dd MMM yyyy HH:mm:ss zzz", Locale.US);
+    static final org.slf4j.Logger LOG = LoggerFactory.getLogger(MethodHandles
+            .lookup().lookupClass());
+
+    public FileResponse(String u, Metadata md, FileProtocol fp)
+            throws MalformedURLException, IOException {
+
+        fileProtocol = fp;
+        metadata = md;
+
+        URL url = new URL(u);
+
+        if (!url.getPath().equals(url.getFile())) {
+            LOG.warn("url.getPath() != url.getFile(): {}.", url);
+        }
+
+        String path = "".equals(url.getPath()) ? "/" : url.getPath();
+
+        File file = new File(
+                URLDecoder.decode(path, fileProtocol.getEncoding()));
+
+        if (!file.exists()) {
+            statusCode = HttpStatus.SC_NOT_FOUND;
+            return;
+        }
+
+        if (!file.canRead()) {
+            statusCode = HttpStatus.SC_UNAUTHORIZED;
+            return;
+        }
+
+        if (!file.equals(file.getCanonicalFile())) {
+            metadata.setValue(HttpHeaders.LOCATION, file.getCanonicalFile()
+                    .toURI().toURL().toString());
+            statusCode = HttpStatus.SC_MULTIPLE_CHOICES;
+            return;
+        }
+
+        if (file.isDirectory()) {
+            getDirAsHttpResponse(file);
+        } else if (file.isFile()) {
+            getFileAsHttpResponse(file);
+        } else {
+            statusCode = HttpStatus.SC_INTERNAL_SERVER_ERROR;
+            return;
+        }
+
+        if (content == null) {
+            content = new byte[0];
+        }
+    }
+
+    public ProtocolResponse toProtocolResponse() throws IOException {
+        return new ProtocolResponse(content, statusCode, metadata);
+    }
+
+    private void getFileAsHttpResponse(File file) throws IOException {
+        long size = file.length();
+
+        if (size > Integer.MAX_VALUE) {
+            statusCode = HttpStatus.SC_BAD_REQUEST;
+            return;
+        }
+
+        int len = (int) size;
+
+        if (fileProtocol.getMaxContentLength() >= 0

```

Comment:
nitpick -> instead of ` String errorMessage` just do

`LOG.error("Exception while fetching file response {} ",file.getPath() , e);`

as that string is not used anywhere else